### PR TITLE
`vit` + `wav2vec2` support

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Adds `ViT` family to BetterTransformer integration
Adds also `Wav2vec2` family to BT integration

Replaces #469 

cc @fxmarty @michaelbenayoun

### CODE REVIEW

File: optimum/bettertransformer/models/encoder_models.py [hunk 1]
```diff
@@ -547,3 +547,211 @@ def forward(self, hidden_states, attention_mask, *_, **__):
         if hidden_states.is_nested and self.is_last_layer:
             hidden_states = hidden_states.to_padded_tensor(0.0)
         return (hidden_states,)
+
+
+class ViTLayerBetterTransformer(BetterTransformerBaseLayer):
+    r"""
```
Comment (fxmarty) [inline optimum/bettertransformer/models/encoder_models.py:553]: Should be moved under the `__init__`. Same issue for existing `WhisperEncoderLayerBetterTransformer`

File: optimum/bettertransformer/models/encoder_models.py [hunk 2]
```diff
@@ -547,3 +547,211 @@ def forward(self, hidden_states, attention_mask, *_, **__):
...
+
+        self.validate_bettertransformer()
+
+    def forward(self, hidden_states, *_, **__):
+        r"""
+        This is just a wrapper around the forward function proposed in:
+        https://github.com/huggingface/transformers/pull/19553
+        """
+        super().forward_checker()
+        # attention mask seems to be always None: https://github.com/huggingface/transformers/blob/94b3f544a1f5e04b78d87a2ae32a7ac252e22e31/src/transformers/models/whisper/modeling_whisper.py#L690
```
Comment (fxmarty) [inline optimum/bettertransformer/models/encoder_models.py:621]: whisper comment in vit, remove?

File: optimum/bettertransformer/models/encoder_models.py [hunk 3]
```diff
@@ -547,3 +547,211 @@ def forward(self, hidden_states, attention_mask, *_, **__):
...
+            self.linear2_bias,
+            attention_mask,
+        )
+        if hidden_states.is_nested and self.is_last_layer:
+            hidden_states = hidden_states.to_padded_tensor(0.0)
+        return (hidden_states,)
+
+
+class Wav2Vec2EncoderLayerBetterTransformer(BetterTransformerBaseLayer):
+    r"""
```
Comment (fxmarty) [inline optimum/bettertransformer/models/encoder_models.py:651]: move in the init

File: optimum/bettertransformer/transformation.py [hunk 4]
```diff
@@ -157,7 +157,17 @@ def transform(
...
-            model_fast = deepcopy(model)
+            model = model.requires_grad_(False)
+            try:
+                model_fast = deepcopy(model)
+            except RuntimeError:
+                raise ValueError(
+                    f"The model {model.__class__.__name__} does not support `deepcopy` operation that is "
+                    " internally used to create a copy of the original model when using"
+                    " `keep_original_model=True`. Please run the conversion with "
+                    " `keep_original_model=Fale` and create a new copy of the original"
```
Comment (fxmarty) [inline optimum/bettertransformer/transformation.py:168]:
```suggestion
                    " `keep_original_model=False` and create a new copy of the original"
```

So you found the issue in the end? For wav2vec2 we can't use `keep_original_model=True`?

File: tests/bettertransformer/test_bettertransformer_audio.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+from transformers import AutoFeatureExtractor, AutoModel, AutoProcessor
+
+from optimum.bettertransformer import BetterTransformer
+from testing_bettertransformer_utils import BetterTransformersTestMixin
+
+
+ALL_AUDIO_MODELS_TO_TEST = [
+    "openai/whisper-tiny",
+    "patrickvonplaten/wav2vec2_tiny_random",
+    "facebook/hubert-base-ls960",
```
Comment (fxmarty) [inline tests/bettertransformer/test_bettertransformer_audio.py:28]: `facebook/hubert-base-ls960` is 350 MB, can you either upload a random tiny hubert to the Hub, or initialize the model in the test?

File: optimum/bettertransformer/models/encoder_models.py (hunk 1 above)
Comment (younesbelkada) [reply optimum/bettertransformer/models/encoder_models.py]: Fixed in  ad02275

File: optimum/bettertransformer/models/encoder_models.py (hunk 3 above)
Comment (younesbelkada) [reply optimum/bettertransformer/models/encoder_models.py]: Fixed in  ad02275

File: optimum/bettertransformer/models/encoder_models.py (hunk 2 above)
Comment (younesbelkada) [reply optimum/bettertransformer/models/encoder_models.py]: Fixed in  ad02275

File: optimum/bettertransformer/transformation.py (hunk 4 above)
Comment (younesbelkada) [reply optimum/bettertransformer/transformation.py:168]: The issue seems to be related to `deepcopy` I didn't managed to fix this yet :/

File: tests/bettertransformer/test_bettertransformer_audio.py (hunk 5 above)
Comment (younesbelkada) [reply tests/bettertransformer/test_bettertransformer_audio.py]: Fixed in 4aff29b

File: optimum/bettertransformer/transformation.py (hunk 4 above)
Comment (younesbelkada) [reply optimum/bettertransformer/transformation.py]: Let's maybe keep this for now, the error message seems quite explicit

File: tests/bettertransformer/test_bettertransformer_vision.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+import requests
+from testing_bettertransformer_utils import BetterTransformersTestMixin
+
+
+ALL_AUDIO_MODELS_TO_TEST = [
+    "hf-internal-testing/tiny-random-ViTModel",
+    "hf-internal-testing/tiny-random-YolosModel",
+    "hf-internal-testing/tiny-random-ViTMAEModel",
+    "hf-internal-testing/tiny-random-ViTMSNModel",
```
Comment (fxmarty) [inline tests/bettertransformer/test_bettertransformer_vision.py:28]:
```suggestion
    "hf-internal-testing/tiny-random-ViTMSNModel",
    "hf-internal-testing/tiny-random-deit",
```

**Review by fxmarty — APPROVED** [review]

File: optimum/bettertransformer/transformation.py (hunk 4 above)
Comment (fxmarty) [reply optimum/bettertransformer/transformation.py]: all good

File: tests/bettertransformer/test_bettertransformer_audio.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+ALL_AUDIO_MODELS_TO_TEST = [
+    "openai/whisper-tiny",
+    "patrickvonplaten/wav2vec2_tiny_random",
+    "ybelkada/hubert-tiny-random",
+]
+
+
+class BetterTransformersWhisperTest(BetterTransformersTestMixin, unittest.TestCase):
+    r""" """
```
Comment (michaelbenayoun) [inline tests/bettertransformer/test_bettertransformer_audio.py:33]: Docstring intended?

File: tests/bettertransformer/test_bettertransformer_audio.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                        hf_random_model.__class__.__name__
+                    ),
+                )
+
+    def test_raise_autocast(self):
+        r"""
+        Autocast seems to be incompatible with `Wav2Vec2` and `cpu`, let's skip this
+        test. The test below should be fine to test if the `forward_checker` has been run correctly.
+        """
+        pass
```
Comment (michaelbenayoun) [inline tests/bettertransformer/test_bettertransformer_audio.py:128]: Why?

**Review by michaelbenayoun — APPROVED** [review]
LGTM!

File: tests/bettertransformer/test_bettertransformer_audio.py (hunk 8 above)
Comment (younesbelkada) [reply tests/bettertransformer/test_bettertransformer_audio.py:128]: I missed somthing there, I think that it should work now

Comment (younesbelkada) [reply tests/bettertransformer/test_bettertransformer_audio.py]: Fixed in [a958177](https://github.com/huggingface/optimum/pull/470/commits/a958177cb8f193ac54e8174fab997a31b85d3750)

File: tests/bettertransformer/test_bettertransformer_audio.py (hunk 7 above)
Comment (younesbelkada) [reply tests/bettertransformer/test_bettertransformer_audio.py]: Fixed in [a958177](https://github.com/huggingface/optimum/pull/470/commits/a958177cb8f193ac54e8174fab997a31b85d3750)

**Pull request merged**